### PR TITLE
[Feature][Log] Integrate slf4j and log4j2 for unified management logs

### PR DESCRIPTION
**Pull request by hailin0**
## Purpose of this pull request

proposal #2725

Change logs:

Maven dependencys:
- Add slf4j & log4j2 provider & bridges scope into root pom.xml dependencyManagement
  - include: slf4j-api、log4j2、log4j2-slf4j-impl(slf4j provider)、xxx-to-slf4j
  - exclude(provided): logback、log4j1、commons-logging、slf4j-other-provider、log4j2-to-slf4j
- Add slf4j & log4j2 & jcl-over-slf4j into root pom.xml dependencies
- Exclude all logging system packages in maven-sheade-plugin

SeaTunnel engine module:
- Enhanced logging http endpoint to dynamically change log4j2 level
- Config `hazelcast.logging.type: log4j2`

SeaTunnel Dist module:
- Use `maven-assembly-plugin` to copy slf4j & log4j2 & jcl-over-slf4j packages to `lib/logging` for use by starter shell

Engine Starter module[flink、spark]:
- Use `maven-dependency-plugin` to copy slf4j & log4j2 & jcl-over-slf4j packages to `target/logging-e2e` for use by E2E testcase
- Add `lib/logging` directory to java classpath in start shell for use by starter class
- Add jcl-over-slf4j package into shade jar
- Config `config/log4j2-console.properties` on engine-client
- Exclude log4j & commons-logging from shade jar
- Fix shell script tail last line text

Engine Starter module[seatunnel]:
- Add `disruptor` jar and config `-Dlog4j2.contextSelector` to use asynchronous logger on engine-server
- Config `config/log4j2-file.properties` on engine-server
- Config `config/log4j2-console.properties` on engine-client
- Add slf4j & log4j2 & jcl-over-slf4j package into starter shade jar
- Support create new log file for submit local mode job

Connector-v1 module[seatunnel-connector-flink-clickhouse]:
- Exclude commons-logging from shade jar

E2E:
- Remove `log4j.properties` files
- Add `log4j2-test.properties` output log to `STDOUT` or `STDERR`
- Copy starter[flink、spark] `target/logging-e2e` into engine container
- Copy config dir into engine container
- Improve execute job logs

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

**Comment by hailin0** [comment]
Add notes to v1 documents：Need to add `runtime` jar to flink engine at runtime

image

File: seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2-test.properties [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+appender.consoleWarn.name = consoleWarn
+appender.consoleWarn.type = CONSOLE
+appender.consoleWarn.target = SYSTEM_ERR
+appender.consoleWarn.layout.type = PatternLayout
+appender.consoleWarn.layout.pattern = %d{yyyy-MM-dd HH:mm:ss,SSS} %-5p %c - %m%n
+appender.consoleWarn.filter.acceptGteWarn.type = ThresholdFilter
+appender.consoleWarn.filter.acceptGteWarn.level = WARN
+appender.consoleWarn.filter.acceptGteWarn.onMatch = ACCEPT
+appender.consoleWarn.filter.acceptGteWarn.onMismatch = DENY
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2-test.properties]: log example:  STDOUT、STDERR

image

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/log/Log4j2HttpGetCommandProcessor.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    @Override
+    public void handleRejection(HttpGetCommand request) {
+        handle(request);
+    }
+
+    @Override
+    public void handle(HttpGetCommand request) {
+        String uri = request.getURI();
+        if (uri.startsWith(HttpCommandProcessor.URI_LOG_LEVEL)) {
+            outputAllLoggerLevel(request);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/log/Log4j2HttpGetCommandProcessor.java:54]: example

image

File: seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/log/Log4j2HttpPostCommandProcessor.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    @Override
+    public void handleRejection(HttpPostCommand request) {
+        handle(request);
+    }
+
+    @Override
+    public void handle(HttpPostCommand request) {
+        String uri = request.getURI();
+        if (uri.startsWith(HttpCommandProcessor.URI_LOG_LEVEL)) {
+            setLoggerLevel(request);
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/log/Log4j2HttpPostCommandProcessor.java:54]: example

image
image

File: seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2-test.properties (hunk 1 above)
Comment (hailin0) [reply seatunnel-e2e/seatunnel-e2e-common/src/test/resources/log4j2-test.properties]: image

**Comment by hailin0** [comment]
network timeout

image

File: config/log4j2-file.properties [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+property.file_path = logs
+property.file_name = seatunnel
+property.file_split_size = 100MB
+property.file_count = 100
+property.file_ttl = 7d
```
Comment (hailin0) [inline config/log4j2-file.properties]: log files：
max bytes = 100MB * 100
max times = 7d

File: seatunnel-core/pom.xml [hunk 5]
```diff
@@ -31,21 +31,31 @@
...
+                    <executions>
+                        <execution>
+                            <id>copy-starter-logging-package-for-e2e</id>
+                            <phase>package</phase>
+                            <goals>
+                                <goal>copy-dependencies</goal>
+                            </goals>
+                            <configuration>
+                                <includeGroupIds>org.slf4j,org.apache.logging.log4j</includeGroupIds>
+                                <outputDirectory>${project.build.directory}/logging-e2e</outputDirectory>
```
Comment (hailin0) [inline seatunnel-core/pom.xml:61]: use by e2e

File: seatunnel-core/pom.xml [hunk 6]
```diff
@@ -31,21 +31,31 @@
...
-                    <shadedArtifactAttached>false</shadedArtifactAttached>
-                    <createDependencyReducedPom>true</createDependencyReducedPom>
-                    <!-- Make sure the transitive dependencies are written to the generated pom under <dependencies> -->
-                    <promoteTransitiveDependencies>true</promoteTransitiveDependencies>
-                    <artifactSet>
-                        <excludes>
-                            <exclude>ch.qos.logback:*</exclude>
-                        </excludes>
-                    </artifactSet>
-                </configuration>
```
Comment (hailin0) [inline seatunnel-core/pom.xml:59]: configuration from extends parent pom

File: seatunnel-core/seatunnel-starter/pom.xml [hunk 7]
```diff
@@ -52,10 +52,45 @@
...
+                            <exclude>org.slf4j:slf4j-reload4j</exclude>
+                            <exclude>org.slf4j:slf4j-log4j12</exclude>
+                            <exclude>log4j:*</exclude>
+                            <exclude>commons-logging:*</exclude>
+                            <exclude>ch.qos.logback:*</exclude>
+                            <!--  -->
+                            <exclude>org.apache.logging.log4j:log4j-to-slf4j</exclude>
+                        </excludes>
+                    </artifactSet>
+                </configuration>
```
Comment (hailin0) [inline seatunnel-core/seatunnel-starter/pom.xml]: Override the parent's configuration

include： slf4j-api & log4j2

File: seatunnel-core/seatunnel-starter/src/main/bin/seatunnel-cluster.sh [hunk 8]
```diff
@@ -60,4 +60,14 @@ if [ -z $SEATUNNEL_CONFIG ]; then
     SEATUNNEL_CONFIG=${CONF_DIR}/seatunnel.yaml
 fi
 
-java -Dseatunnel.config=${SEATUNNEL_CONFIG} -Dhazelcast.config=${HAZELCAST_CONFIG} -cp ${APP_JAR} ${APP_MAIN} ${args}
+if [ -z $LOG4J2_CONFIG ]; then
+    LOG4J2_CONFIG=${CONF_DIR}/log4j2-file.properties
+fi
+
+if [ -z $LOG4J2_CONTEXT_SELECTOR ]; then
+    LOG4J2_CONTEXT_SELECTOR="org.apache.logging.log4j.core.async.AsyncLoggerContextSelector"
```
Comment (hailin0) [inline seatunnel-core/seatunnel-starter/src/main/bin/seatunnel-cluster.sh]: Enable async logger

File: seatunnel-engine/seatunnel-engine-common/src/main/resources/hazelcast.yaml [hunk 9]
```diff
@@ -17,7 +17,14 @@
...
 hazelcast:
   cluster-name: seatunnel
+  properties:
+    hazelcast.logging.type: log4j2
   network:
+    rest-api:
+      enabled: true
+      endpoint-groups:
+        CLUSTER_WRITE:
+          enabled: true
```
Comment (hailin0) [inline seatunnel-engine/seatunnel-engine-common/src/main/resources/hazelcast.yaml:25]: Enable API:       POST /hazelcast/rest/log-level

File: seatunnel-core/pom.xml (hunk 5 above)
Comment (hailin0) [reply seatunnel-core/pom.xml:61]: 企业微信截图_c16d0688-371c-4240-a391-8fc404626ef7

File: seatunnel-core/pom.xml [hunk 10]
```diff
@@ -31,18 +31,47 @@
...
+                            <exclude>org.slf4j:slf4j-jdk14</exclude>
+                            <exclude>org.slf4j:slf4j-nop</exclude>
+                            <exclude>org.slf4j:slf4j-simple</exclude>
+                            <exclude>org.slf4j:slf4j-reload4j</exclude>
+                            <exclude>org.slf4j:slf4j-log4j12</exclude>
                             <exclude>ch.qos.logback:*</exclude>
+                            <exclude>log4j:*</exclude>
+                            <exclude>org.apache.logging.log4j:*</exclude>
+                            <exclude>commons-logging:*</exclude>
                         </excludes>
```
Comment (hailin0) [inline seatunnel-core/pom.xml]: starter shade jar include jcl-over-slf4j


image

File: seatunnel-core/seatunnel-starter/pom.xml [hunk 11]
```diff
@@ -52,10 +52,44 @@
...
+                            <exclude>org.slf4j:slf4j-jdk14</exclude>
+                            <exclude>org.slf4j:slf4j-nop</exclude>
+                            <exclude>org.slf4j:slf4j-simple</exclude>
+                            <exclude>org.slf4j:slf4j-reload4j</exclude>
+                            <exclude>org.slf4j:slf4j-log4j12</exclude>
+                            <exclude>log4j:*</exclude>
+                            <exclude>commons-logging:*</exclude>
+                            <exclude>ch.qos.logback:*</exclude>
+                            <exclude>org.apache.logging.log4j:log4j-to-slf4j</exclude>
+                        </excludes>
```
Comment (hailin0) [inline seatunnel-core/seatunnel-starter/pom.xml]: seatunnel starter shade jar include : slf4j-api、log4j-api、log4j-core、log4j-slf4j-impl、jcl-over-slf4j、disruptor

image

File: seatunnel-dist/src/main/assembly/assembly-bin-ci.xml [hunk 12]
```diff
@@ -115,6 +115,21 @@
...
+            <useTransitiveDependencies>true</useTransitiveDependencies>
+            <unpack>false</unpack>
+            <outputDirectory>/lib/logging</outputDirectory>
+            <includes>
+                <include>org.slf4j:slf4j-api:jar</include>
+                <include>org.slf4j:jcl-over-slf4j:jar</include>
+                <include>org.apache.logging.log4j:log4j-api:jar</include>
+                <include>org.apache.logging.log4j:log4j-core:jar</include>
+                <include>org.apache.logging.log4j:log4j-slf4j-impl:jar</include>
+            </includes>
```
Comment (hailin0) [inline seatunnel-dist/src/main/assembly/assembly-bin-ci.xml:130]: 企业微信截图_ffd01110-a84e-46c4-b1d8-4ab07c730025

File: seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java [hunk 13]
```diff
@@ -83,6 +83,17 @@ public static String copyConfigFileToContainer(GenericContainer<?> container, St
...
+                                                     String startModulePath,
+                                                     String seatunnelHomeInContainer) {
+        // copy logging lib
+        final String loggingLibPath = startModulePath + File.separator + "target" + File.separator + "logging-e2e" + File.separator;
+        checkPathExist(loggingLibPath);
+        container.copyFileToContainer(
+            MountableFile.forHostPath(loggingLibPath),
+            Paths.get(seatunnelHomeInContainer, "lib", "logging").toString());
+    }
+
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-e2e-common/src/test/java/org/apache/seatunnel/e2e/common/util/ContainerUtil.java:104]: mapping module dir to Container:

module dir:
企业微信截图_c16d0688-371c-4240-a391-8fc404626ef7

Container dir:
image

**Comment by hailin0** [comment]
Improve maven package：

**starter[spark、flink]**
change before:
starter shade jar include: slf4j、 log4j1.x
企业微信截图_bf7a05da-21c3-4fb6-996c-8cd9bb29a144


change after:
starter shade jar exclude: slf4j、 log4j1.x,   include:jcl-over-slf4j
企业微信截图_f5f9b718-7404-49f1-a47c-d35d2e68c970


**connector：example seatunnel-connector-flink-clickhouse**
change before:
connector shade jar include: commons-logging, or other loggings
企业微信截图_574e6a49-c39a-45f7-a744-cb1fc068e6eb

change after:
connector shade jar exclude all logging packages
企业微信截图_1d2163d6-351f-480f-a02e-73660793dbc8

**Comment by EricJoy2048** [comment]
Please fix CI problem.

**Comment by hailin0** [comment]
> Please fix CI problem.

image

File: config/log4j2.properties [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+property.file_path = ${sys:seatunnel.logs.path:-/tmp/seatunnel/logs}
+property.file_name = ${sys:seatunnel.logs.file_name:-seatunnel}
+property.file_split_size = 100MB
+property.file_count = 100
+property.file_ttl = 7d
```
Comment (hailin0) [inline config/log4j2.properties:23]: log files：
max bytes = 100MB * 100
max times = 7d

File: config/log4j2.properties [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+rootLogger.level = INFO
+
+################################# console log #################################
+rootLogger.appenderRef.consoleInfo.ref = consoleInfo
+rootLogger.appenderRef.consoleWarn.ref = consoleWarn
+################################# console log #################################
+################################# file log    #################################
+#rootLogger.appenderRef.file.ref = fileAppender
```
Comment (hailin0) [inline config/log4j2.properties:32]: Disable file log

File: seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh [hunk 16]
```diff
@@ -50,15 +50,24 @@ else
     args=$@
 fi
 
+set +u
+# Log4j2 Config
+if [ -e "${CONF_DIR}/log4j2.properties" ]; then
+  JAVA_OPTS="${JAVA_OPTS} -Dlog4j2.configurationFile=${CONF_DIR}/log4j2.properties"
+  JAVA_OPTS="${JAVA_OPTS} -Dseatunnel.logs.path=${APP_DIR}/logs"
+  JAVA_OPTS="${JAVA_OPTS} -Dseatunnel.logs.file_name=seatunnel-core-flink-sql"
```
Comment (hailin0) [inline seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh:58]: Add log4j2 system properties

File: seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh [hunk 17]
```diff
@@ -50,15 +50,24 @@ else
...
+fi
+
+CLASS_PATH=${APP_DIR}/lib/logging/*:${APP_JAR}
```
Comment (hailin0) [inline seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh]: add `lib/logging/` into starter jar classpath

File: seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh [hunk 18]
```diff
@@ -50,15 +50,24 @@ else
...
+CMD=$(java ${JAVA_OPTS} -cp ${CLASS_PATH} ${APP_MAIN} ${args}) && EXIT_CODE=$? || EXIT_CODE=$?
 if [ ${EXIT_CODE} -eq 234 ]; then
     # print usage
     echo "${CMD}"
     exit 0
 elif [ ${EXIT_CODE} -eq 0 ]; then
-    echo "Execute SeaTunnel Flink SQL Job: ${CMD}"
-    eval ${CMD}
+    echo "Execute SeaTunnel Flink SQL Job: $(echo "${CMD}" | tail -n 1)"
+    eval $(echo "${CMD}" | tail -n 1)
```
Comment (hailin0) [inline seatunnel-core/seatunnel-core-flink-sql/src/main/bin/start-seatunnel-sql.sh:70]: tail last line text

File: seatunnel-core/seatunnel-core-spark/src/main/bin/start-seatunnel-spark.sh [hunk 19]
```diff
@@ -49,14 +49,24 @@ else
...
+CMD=$(java ${JAVA_OPTS} -cp ${CLASS_PATH} ${APP_MAIN} ${args}) && EXIT_CODE=$? || EXIT_CODE=$?
 if [ ${EXIT_CODE} -eq 234 ]; then
     # print usage
     echo "${CMD}"
     exit 0
 elif [ ${EXIT_CODE} -eq 0 ]; then
-    echo "Execute SeaTunnel Spark Job: $(echo ${CMD} | tail -n 1)"
-    eval $(echo ${CMD} | tail -n 1)
+    echo "Execute SeaTunnel Spark Job: $(echo "${CMD}" | tail -n 1)"
+    eval $(echo "${CMD}" | tail -n 1)
```
Comment (hailin0) [inline seatunnel-core/seatunnel-core-spark/src/main/bin/start-seatunnel-spark.sh:69]: Fix tail last line text: ${CMD} -> "${CMD}"

**Comment by hailin0** [comment]
@TyrantLucifer  @EricJoy2048 @ashulin @Hisoka-X @CalvinKirs   thx. PTAL

**Comment by hailin0** [comment]
close #2474

**Comment by EricJoy2048** [comment]
resolve conflicts please.

**Review by ashulin — COMMENTED** [review]
@CalvinKirs CC, Does he need to resolve license conflicts?

File: seatunnel-core/seatunnel-starter/pom.xml [hunk 20]
```diff
@@ -52,10 +52,44 @@
...
+                    </artifactSet>
+                </configuration>
+            </plugin>
+            <!-- make sure that flatten runs after maven-shade-plugin -->
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>flatten-maven-plugin</artifactId>
+            </plugin>
```
Comment (hailin0) [inline seatunnel-core/seatunnel-starter/pom.xml]: Do I need to delete?
@ashulin 

https://github.com/apache/incubator-seatunnel/pull/3078

**Review by CalvinKirs — COMMENTED** [review]
image
These two can be removed and we do not need to declare the standard AL2

**Review by EricJoy2048 — DISMISSED** [review]
+1

**Review by CalvinKirs — APPROVED** [review]

**Comment by hailin0** [comment]
image

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**